src/Java.Interop.Tools.JavaTypeSystem/Adapters/JavaXmlApiImporter.cs

```diff
@@ -0,0 +1,370 @@
+			// That is, 'android.foo.Blah' needs to be added before 'android.foo.Blah.Bar'.
+			foreach (var type in packages.SelectMany (p => p.Types).Where (t => t.NestedName.Contains ('.')).OrderBy (t => t.FullName.Count (c => c == '.')).ToArray ()) {
+				collection.AddType (type);
+
+				// Remove nested types from Package
+				type.Package.Types.Remove (type);
+			}
+
+			return collection;
+		}
+
+		public static JavaPackage ParsePackage (XElement package, JavaTypeCollection collection)
+		{
+			var pkg = collection.AddPackage (
+				name: package.XGetAttribute ("name"),
+				jniName: package.XGetAttribute ("jni-name"),
+				managedName: package.XGetAttributeOrNull ("managedName")
+			);
+
+			if (package.XGetAttribute ("merge.SourceFile") is string source && source.HasValue ())
+				pkg.PropertyBag.Add ("merge.SourceFile", source);
+
+			foreach (var elem in package.Elements ()) {
+				switch (elem.Name.LocalName) {
+					case "class":
+						if (package.XGetAttributeAsBool ("obfuscated"))
+							continue;
+
+						pkg.Types.Add (ParseClass (pkg, elem));
+						break;
+					case "interface":
+						if (package.XGetAttributeAsBool ("obfuscated"))
+							continue;
+
+						pkg.Types.Add (ParseInterface (pkg, elem));
+						break;
+				}
+			}
+
+			return pkg;
+		}
+
+		public static JavaClassModel ParseClass (JavaPackage package, XElement element)
+		{
+			var model = new JavaClassModel (
+				javaPackage: package,
+				javaNestedName: element.XGetAttribute ("name"),
+				javaVisibility: element.XGetAttribute ("visibility"),
+				javaAbstract: element.XGetAttributeAsBool ("abstract"),
+				javaFinal: element.XGetAttributeAsBool ("final"),
+				javaBaseType: element.XGetAttribute ("extends"),
+				javaBaseTypeGeneric: element.XGetAttribute ("extends-generic-aware"),
+				javaDeprecated: element.XGetAttribute ("deprecated"),
+				javaStatic: element.XGetAttributeAsBool ("static"),
+				jniSignature: element.XGetAttribute ("jni-signature"),
+				baseTypeJni: element.XGetAttribute ("jni-extends")
+			);
+
+			if (element.XGetAttribute ("merge.SourceFile") is string source && source.HasValue ())
+				model.PropertyBag.Add ("merge.SourceFile", source);
+			if (element.XGetAttribute ("deprecated-since") is string dep && dep.HasValue ())
+				model.PropertyBag.Add ("deprecated-since", dep);
+
+			if (element.Element ("typeParameters") is XElement tp)
+				ParseTypeParameters (model.TypeParameters, tp);
+
+			foreach (var child in element.Elements ()) {
+				switch (child.Name.LocalName) {
+					case "constructor":
+						model.Constructors.Add (ParseConstructor (model, child));
+						break;
+					case "field":
+						model.Fields.Add (ParseField (model, child));
+						break;
+					case "implements":
+						model.Implements.Add (ParseImplements (child));
+						break;
+					case "method":
+						model.Methods.Add (ParseMethod (model, child));
+						break;
+				}
```

Comment:
Should we warn on any other unexpected element values?

Reply:
Same reply about generating too many warnings.